Common/include/toolboxes/CModule.hpp

```diff
@@ -0,0 +1,48 @@
+#include <tuple>
+#include <utility>
+#pragma once
+
+
+#define CREATE_ACTION(name, func_args...) \
+  static struct Action##name {} Action##name; \
+  template <typename type, typename... Targs> \
+  static void Action(type& Module, struct Action##name, Targs& ...args){ Module.name(args...); }
+
+
+template<class DerivedModule>
+class CModule {
+
+protected:
+
+  bool enabled;
+
+public:
+
+  CModule(bool enabled_ = true) :
+    enabled(enabled_){}
+
+  bool IsEnabled() { return enabled; }
+
+  template<std::size_t I = 0, typename TAction, typename... Tp, typename... Targs>
+  inline static typename std::enable_if<I == sizeof...(Tp), void>::type
+  for_each(std::tuple<Tp...> &, TAction, Targs&...) { }
```

Comment:
I opted for a compile-time construction of the iteration over the module hooks. But that is up or discussion. Can be very well also be done dynamically.